### ISSUE TITLE
Handle IntegrityError from race conditions

### ISSUE THREAD

**Issue by jace** [hasgeek/coaster#58]
Every once in a while, a user manages to double click a submit button, causing their browser to send two identical requests, causing our web server to fire up two parallel workers to handle the request (workers A & B), which then perform this dance:

* A checks for an existing record, finds none
* B checks for an existing record, finds none
* A makes a new record and attempts to commit
* B makes a new record and attempts to commit
* A's commit succeeds
* B's commit fails with an IntegrityError (some unique constraint was violated)
* A returns status to browser, but browser has since cancelled the request in favour of B
* B returns failure, resulting in user seeing a failure message when the operation actually succeeded

While we have changed our UI to disable submit buttons after the first submit (in hasgeek/baseframe@b03b666fd5b1052f62b583d2441d4de6a8736dff), the correct solution is to use a SAVEPOINT and roll-back just that failure. This has been documented in hasgeek/lastuser#130, but since it requires 7 lines of code, it's best abstracted into a base class. Here is the [code outline from Mike Bayer](https://groups.google.com/forum/m/#!topic/sqlalchemy/neg7nXagaMQ):

```python
session.begin_nested() 
try: 
     session.add(thing_that_may_exist_already) 
     session.commit()  # flushes, and commits only the "savepoint" 
except exc.IntegrityError: 
    session.rollback()     
    thing_that_may_exist_already = session.query(Thing).filter_by(<criteiron>).one() 
```

In our case, we could have:

```python
class AddAndCommitMixin(object):  # mixin class added to BaseMixin
    @classmethod
    def add_and_commit(cls, ob, loader, *args, **kwargs):
        cls.query.session.begin_nested()
        try:
            cls.query.session.add(ob)
            cls.query.session.commit()
            return ob
        except exc.IntegrityError:
            cls.query.session.rollback()
            return loader(*args, **kwargs)
```

Typical use would be like this (assuming Lastuser's UserEmail confirmation, where the confirmation happens in a GET request via their email client, and so has no protection from disabling submit buttons):

```python
useremail = UserEmail(user=g.user, email=email)
useremail = UserEmail.add_and_commit(useremail, UserEmail.get, user=g.user, email=email)
```

**Comment by jace** [comment]
From an aesthetic standpoint, `add_and_commit` should be a method on the session object, not the model. However:

1. We don't use custom session objects yet, so this is unexplored territory supplying our own (we had a similar issue with query objects when adding `one_or_none()` in 32e1e8d36a6b1edbca99491dc0af309121674514 and decided then it was time for a custom query class)

2. This code assumes the loader returns an instance, depending on our now-standard `get` method pattern. If we move one layer down, we have the SQLAlchemy-provided `query.filter_by(**kwargs)` which returns a query, not a model instance. Since the `add_and_commit` method returns an instance if the commit was successful, it must return an instance even if not. This means it will need to check if the loader returned a query and execute `.one()` on it (not `.one_or_none()` as (a) an instance must exist, and (b) that method may not be available on the given query object).

**Comment by jace** [comment]
Flask-SQLAlchemy requires you to subclass it to replace the default session object (see [here](https://github.com/mitsuhiko/flask-sqlalchemy/blob/cfe529a5997edcfd65fbcc4b4928be180bfa7e7b/flask_sqlalchemy/__init__.py#L767-L773) and [here](https://github.com/mitsuhiko/flask-sqlalchemy/blob/cfe529a5997edcfd65fbcc4b4928be180bfa7e7b/flask_sqlalchemy/__init__.py#L136-L142)). Our options are:

1. Don't bother with the session and place this method in the model.
2. Monkey patch the session class in SQLAlchemy or Flask-SQLAlchemy
3. Subclass Flask-SQLAlchemy's SQLAlchemy's class and use it in all apps (easier with coaster.db as a centralised place, but some apps don't use it yet – maybe they can switch along with the move to `add_and_commit`)

**Comment by jace** [comment]
Initial code is in 5f0de22.

**Comment by jace** [comment]
Pull request in #60. Review please, @shreyas-satish.

**Comment by jace** [comment]
Fixed in #60.
